docs/RFCS/sql_column_groups.md

```diff
@@ -0,0 +1,122 @@
+non-primary key column. The key is structured as:
+
+```
+/<tableID>/<indexID>/<primaryKeyColumns...>/<columnID>
+```
+
+Keys also have an additional suffix imposed by the MVCC
+implementation: ~10 bytes of timestamp.
+
+The value is a proto (note, this is what is stored on disk, we also
+include a timestamp when transmitting over the network):
+
+```
+message Value {
+  optional bytes raw_bytes;
+  optional fixed32 checksum;
+  optional ValueType tag;
+}
+```
+
+Consider the following table:
+
+```
+CREATE TABLE kv (
+  k INT PRIMARY KEY,
+  v INT
+)
+```
+
+A row for this table logically has 16 bytes of data (`INTs` are 64-bit
+values). The current SQL to KV mapping will create 2 key/value pairs:
+a sentinel key for the row and the key/value for column `v`. These two
+keys imply an overhead of ~28 bytes (timestamp + checksum) when stored
+on disk and more than that when transmitted over the network (disk
+storage takes advantage of prefix compression of the keys). We can cut
+this overhead in half by storing only a single key/value for the
+row. The savings are even more significant as more columns are added
+to the table.
+
+Note that there is also per key overhead at the transaction
+level. Every key written within a transaction has a "write intent"
+associated with it and these intents need to be resolved when the
+transaction is committed. Reducing the number of keys per row reduces
+this overhead.
+
+# Detailed design
+
+The structure for table keys will be changed to:
+
+```
+/<tableID>/<indexID>/<primaryKeyColumns...>/<groupID>
+```
+
+The value proto will remain almost the same, but the `tag` field will
+be removed and instead of encoding the column value directly into the
+`raw_bytes` field we'll encoding a series of `<columnID>/<columnVal>`
+pairs where `<columnID>` is varint encoded and `<columnVal>` is
+encoded per its type using the `util/encoding` routines. Similar to
+the existing convention, `NULL` column values will not be present in
+the value.
+
+The `ColumnDescriptor` proto will be extended with a `GroupID`
+field. `GroupIDs` will be allocated per-table using a new
+`TableDescriptor.NextGroupID` field. `GroupID(0)` will be always be
+written for a row and will take the place of the sentinel key. We
+could even omit the `/<groupID>` key suffix when encoding `GroupID(0)`
+to make the encoding identical to the previous sentinel key encoding.
+
+When a column is added to a table it will be assigned to a group. A
+group will be selected so that raw column data size for the columns in
+the group is less than some fixed size (e.g. 256 bytes). Note that
+assigning every column to its own group would be effectively
+equivalent to the current SQL to KV mapping.
+
+# Drawbacks
+
+* Grouping multiple columns into a single key/value pair will
+necessitate changes to scan, insert, update and delete. Update will
+require special attention as the `NULL`-ing of a column will be
+performed via a `Put` instead of a `Del`.
```

Comment:
The read followed by a write is already covered by the existing scan done at the beginning of any UPDATE or DELETE. New KV level commands almost certainly make sense at some point. Independent of this change, I'd like for a Scan operation to only return the columns required by the query.